File: leetcode/cpp/3258.count-substrings-that-satisfy-k-constraint-i.cpp
class Solution {
public:
    int countKConstraintSubstrings(string s, int k) {
        int ans=0, zeros=0, ones=0, left=0, right=0;
        for (right=0; right < s.length(); right++) {
            if (s[right] == '0') {
                zeros++;
            } else {
                ones++;
            }
            while (zeros > k && ones > k) {
                if (s[left] == '0') {
                    zeros--;
                } else {
                    ones--;
                }
                left++;
            }
            ans += (right - left + 1);
        }
        return ans;
    }
};

File: leetcode/go/3258.count-substrings-that-satisfy-k-constraint-i.go
func countKConstraintSubstrings(s string, k int) int {
    ans, zeros, ones, left, right := 0, 0, 0, 0, 0
    for right = 0; right < len(s); right++ {
        if s[right:right+1] == "0" {
            zeros++
        } else {
            ones++
        }
        for left < right && zeros > k && ones > k {
            if s[left:left+1] == "0" {
                zeros--
            } else {
                ones--
            }
            left++
        }
        ans += (right - left + 1)
    }
    return ans
}

